### ISSUE TITLE
`Set default country from browser language` doesn't work for shipping carriers

### ISSUE THREAD

**Issue by Traumflug** [thirtybees/thirtybees#991]
Hello, for now I avoid to open a new "issue": please, why "`Set default country from browser language`" deesn't work for shipping carriers?
That is, if the client browser is in german, IF the client is **not logged in** (visitor) just select the carriers for Germany and so on (even if other countries as Austria use german, no problem)...but why it doesn't work at the moment?
Thanks

_Originally posted by @7starsone in https://github.com/thirtybees/thirtybees/issues/977#issuecomment-508912384_

**Comment by 7starsone** [comment]
I think that's here on Carrier class:
`public static function getAvailableCarrierList(`

at the beginning instead of a check on the default country by browser (for the client)...
it does
```
if ($psCountryDefault === null) {
            $psCountryDefault = Configuration::get('PS_COUNTRY_DEFAULT');
        }
``` 
so, it just gets the set (and fixed) default country on database....

and then...
```
if ($idAddress) {
...
} else {
            $country = new Country($psCountryDefault);
            $idZone = $country->id_zone;
        }
```
so, for not logged in users, it does nothing but get the carriers for the **fixed default country**.

Now, sorry, unfortunately I have no time to fix it, would you like to get a look at it, please?
Thanks

**Comment by 7starsone** [comment]
well, `getAvailableCarrierList` is not enough...
so, to be sure you should override `getAvailableCarrierList`, `getDeliveryPriceByWeight` and `checkDeliveryPriceByWeight`, at least.

Simple example for language if you have a few languages and countries enabled (instead of browser/geolocation):
```
public static function checkDeliveryPriceByWeight($id_carrier, $total_weight, $id_zone)
    {
        $logged = Context::getContext()->customer->isLogged();
            //at first you check if customer is not logged, so you set the id_zone by language (check it on DB ps_ country)
            if ($logged == false) {
            $context = Context::getContext();
            $id_lang = $context->language->id;

            if ($id_lang == 1) $id_zone = x;
            if ($id_lang == 2) $id_zone = xx;
            if ($id_lang == 3) $id_zone = xx;
            // and so on...
                  }
...
```
this is just a workaround for specific needs, for a complete and global solution it needs check for browser language/geolocation, a foreach and so on... the rest is up to you.

**Comment by getdatakick** [comment]
Thank you for investigation.

I agree that the system should take browser language/geolocation into account when resolving carrier for non-logged users. But I don't think these methods you mentioned should do that. It's because they can be used from different context (back office, front office, webservices,...), and geolocation makes sense only for front-end usage.

Instead, these methods should be refactored, and receive all required information by input parameters. Then we need to fix the all places in *front office* calling these methods and provide country information from there (using browser language / gelolocation / or any other means).

**Comment by 7starsone** [comment]
@getdatakick  yes, sure, you're right. Anyway at the moment if you just need to check the context for `customer` front-end, the suggested edit works fine

**Comment by Traumflug** [comment]
Marking as Enhancement, because lack of a feature, not malfunction of existing code.

**Comment by Dh42** [comment]
One thing we have to be careful with is showing prices. I had a merchant this week that got their google shopping account suspended for something stupid like this. I am not sure that Google uses a default country with googlebot, but if you show the wrong shipping prices they will suspend your account.

**Comment by 7starsone** [comment]
@Dh42 exactly, at first you should consider that you send the shipping prices to Google by feed, so when adding to cart prices should match even if product price is more important for sure. I think that for some pages if you set auto update on merchant center maybe is negligible but if many product/shipping prices are wrong the problem is there.

**Comment by 7starsone** [comment]
Speaking of Google, unless I'm mistaken you can send just one carrier price list...but, anyway, it should match on the checkout (among the others), at least 😄 I don't think Google is too restrictive about this, but if no price of no carrier matches...mmmm mmmm some troubles expected...

**Comment by Dh42** [comment]
Now they want the carriers for all countries and you cannot have shipping options that you do not send over to them. That is what happened to my client, they had special shipping options that were not sent to google. Also, you HAVE to have guest checkout enabled too... its becoming a PITA.

**Comment by 7starsone** [comment]
"`guest checkout enabled`" honestly, in some countries it breaks the law...in some countries you need complete data from the customer, eg. above all if enterprise, and also many merchants not even register credit cart data and so on...so this seems extremely doubtful that a private company can force you for a guest checkout

**Comment by Dh42** [comment]
The only difference in a guest checkout and a regular checkout is a password.
